File: ImageFlow/ocaml/bridge/objc.c
#include <objc/objc.h>
#include <objc/objc-runtime.h>

#include <caml/mlvalues.h>
#include <caml/memory.h>
#include <caml/alloc.h>
#include <caml/custom.h>

#include "objc.h"

static void objc_finalize(value v) {
  static SEL release = NULL;
  if (release == NULL)
    release = sel_registerName("release");
  objc_msgSend(objc_unwrap(v), release);
}

static long objc_hash(value v) {
  CAMLparam1(v);
  static SEL hash = NULL;
  if (hash == NULL)
    hash = sel_registerName("hash");
  CAMLreturn((long)objc_msgSend(objc_unwrap(v), hash));
}

static struct custom_operations objc_custom_ops = {
  "com.imageflow.objc",
  objc_finalize,
  custom_compare_default,
  objc_hash,
  custom_serialize_default,
  custom_deserialize_default
};

value objc_really_wrap(void* objc_object, int retainFirst) {
  static SEL retain = NULL;
  if (retain == NULL)
    retain = sel_registerName("retain");
  if (retainFirst)
    objc_msgSend(objc_object, retain);
  value block = caml_alloc_custom(&objc_custom_ops, sizeof(void*), 0, 1);
  *((void**)Data_custom_val(block)) = objc_object;
  return block;
}

value objc_wrap(void* objc_object) {
  return objc_really_wrap(objc_object, 1);
}

value objc_wrap_no_retain(void* objc_object) {
  return objc_really_wrap(objc_object, 0);
}

File: ImageFlow/ocaml/bridge/objc.m
#import <caml/mlvalues.h>
#import <caml/memory.h>
#import <caml/custom.h>

#import <Foundation/Foundation.h>

#import "objc.h"

static void objc_finalize(value v) {
  [objc_unwrap(v) release];
}

static long objc_hash(value v) {
  return [objc_unwrap(v) hash];
}

static int objc_compare(value v1, value v2) {
  NSObject* o1 = objc_unwrap(v1);
  NSObject* o2 = objc_unwrap(v2);
  if (o1 == o2 || [o1 isEqual:o2])
    return 0;
  else if (o1 < o2)
    return -1;
  else
    return 1;
}

static struct custom_operations objc_custom_ops = {
  "com.imageflow.objc",
  objc_finalize,
  objc_compare,
  objc_hash,
  custom_serialize_default,
  custom_deserialize_default
};

value objc_really_wrap(id objc_object, BOOL retainFirst) {
  value block = caml_alloc_custom(&objc_custom_ops, sizeof(void*), 0, 1);
  *((id*)Data_custom_val(block)) =
    retainFirst ? [objc_object retain] : objc_object;
  return block;
}

value objc_wrap(id objc_object) {
  return objc_really_wrap(objc_object, YES);
}

value objc_wrap_no_retain(id objc_object) {
  return objc_really_wrap(objc_object, NO);
}
